lib/http-api.js

```diff
@@ -148,16 +155,47 @@ module.exports.MatrixHttpApi.prototype = {
                 filename: file.name,
                 access_token: this.opts.accessToken
             };
-            file.stream.pipe(
-                this.opts.request({
-                    uri: url,
-                    qs: queryParams,
-                    method: "POST"
-                }, requestCallback(defer, callback, this.opts.onlyData))
-            );
+            upload.request = this.opts.request({
+                uri: url,
+                qs: queryParams,
+                method: "POST"
+            }, requestCallback(defer, callback, this.opts.onlyData));
+            file.stream.pipe(this.opts.request);
         }
 
-        return defer.promise;
+        this.uploads.push(upload);
+
+        var self = this;
+        upload.promise = defer.promise.finally(function() {
+            var uploadsKeys = Object.keys(self.uploads);
+            for (var i = 0; i < uploadsKeys.length; ++i) {
+                if (self.uploads[uploadsKeys[i]].promise === defer.promise) {
+                    self.uploads.splice(uploadsKeys[i], 1);
+                }
+            }
+        });
+        return upload.promise;
+    },
+
+    cancelUpload: function(promise) {
+        var uploadsKeys = Object.keys(this.uploads);
+        for (var i = 0; i < uploadsKeys.length; ++i) {
+            var upload = this.uploads[uploadsKeys[i]];
+            if (upload.promise === promise) {
```

Comment:
comparing promise objects seems like a bit of an ugly way to track which upload is being manipulated here, but I guess this is to avoid changing the upload API to return something more useful than a promise. ah well.